File: 448. Find All Numbers Disappeared in an Array.cpp
class Solution {
public:
    // Hashset: O(N) time & space, 2.93%
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        vector<int> ret;
        if(nums.empty()) return ret;
        set<int> flag;
        for(auto num: nums) flag.insert(num);
        for(int i = 1; i <= nums.size(); i++)
            if(!flag.count(i))
                ret.push_back(i);
        return ret;
    }

    // for those already seen numbers, apply the application to the element with these numbers as indexes: negate/+n/...
    // O(N) time, without extra space, 35.82%, ok
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        vector<int> ret;
        if(nums.empty()) return ret;
        for(auto num: nums){
            int index = abs(num) - 1;  // bug: index = num - 1; num may already be negated -> index underflow
            if(nums[index] > 0)
                nums[index] = -nums[index];
        }
        for(int i = 0; i < nums.size(); i++){
            if(nums[i] > 0)
                ret.push_back(i + 1);}
        return ret;
    }
};

File: 448. Find All Numbers Disappeared in an Array.py
class Solution(object):
    
    # sort: O(NlogN) time, no space, 21.03%
    def findDisappearedNumbers(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        nums.sort()
        ret, i = list(), 0
        for j in xrange(1, len(nums)+1):
            if i >= len(nums) or nums[i] > j:
                ret.append(j)
            else:
                while i < len(nums) and nums[i] == j:
                    i += 1
        return ret
            
    # set: O(min(m,n)) time, O(N) space, 77.37%
    def findDisappearedNumbers(self, nums):
        return list(set(nums) ^ set(range(1, len(nums)+1)))
        
    # brilliant: O(N) time, no space, 8.43%
    # for each num in nums, mark the number at index (num-1) as negative
    # after one pass, all index (+1) that are not positive are not visited
    def findDisappearedNumbers(self, nums):
        for num in nums:
            index = abs(num) - 1
            nums[index] = -abs(nums[index])  # bug: avoid double flip 
        return [i+1 for i in xrange(len(nums)) if nums[i] > 0]
            
        